compiler/next/test/resolution/testPoi.cpp

```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/resolution/resolution-queries.h"
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/uast/Call.h"
+#include "chpl/uast/Comment.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Variable.h"
+
+// always check assertions in this test
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <cassert>
+
+using namespace chpl;
+using namespace parsing;
+using namespace resolution;
+using namespace uast;
+
+// test showing point-of-instantiation behavior
+static void test1() {
+  printf("test1\n");
+  Context ctx;
+  Context* context = &ctx;
+
+  auto path = UniqueString::build(context, "input.chpl");
+  std::string contents = "module M {\n"
+                         "  proc generic(param p) {\n"
+                         "    helper();\n"
+                         "  }\n"
+                         "}\n"
+                         "module N {\n"
+                         "  use M;\n"
+                         "  proc helper() { }\n"
+                         "  generic(1);\n"
```

Comment:
I guess the point of this test is to show that we can instantiate generic here (that is on line 56) because helper() is in the scope as of the point-of-instantiation, right? If we tried to add a call to generic() in module M that would fail because helper doesn't exist? 

Maybe it seems obvious, but I think I would explicitly spell that out (in comments) that this is what the test is demonstrating. If you don't already, I think I would also follow up with a negative test showing that we fail to resolve generic() if we instantiate it in a module without helper().

If whatever reason I needed to understand what PoiScope, PoiInfo, etc. are for, I'd probably grep for it, find this unit test, and try and understand what the test is demonstrating.

Reply:
Sure. I'll add a bit more comments and a negative test. The language spec section https://chapel-lang.org/docs/language/spec/generics.html#function-visibility-in-generic-functions talks about this point-of-instantiation behavior. (I have been hesitant to link to the spec from the tests b/c I expect we will rearrange the spec eventually).